README.md
# Deprecated: please visit https://github.com/amundsen-io/amundsen/tree/main/metadata

The Amundsen project [moved to a monorepo](https://github.com/amundsen-io/rfcs/pull/31). This repository will be kept up temporarily to allow users to transition gracefully, but new PRs won't be accepted.

# Amundsen Metadata Service
[![PyPI version](https://badge.fury.io/py/amundsen-metadata.svg)](https://badge.fury.io/py/amundsen-metadata)
[![Build Status](https://api.travis-ci.com/lyft/amundsenmetadatalibrary.svg?branch=master)](https://travis-ci.com/lyft/amundsenmetadatalibrary)
[![Coverage Status](https://img.shields.io/codecov/c/github/lyft/amundsenmetadatalibrary/master.svg)](https://codecov.io/github/lyft/amundsenmetadatalibrary?branch=master)
[![License](https://img.shields.io/:license-Apache%202-blue.svg)](LICENSE)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](#developer-guide)
[![Slack Status](https://img.shields.io/badge/slack-join_chat-white.svg?logo=slack&style=social)](https://amundsenworkspace.slack.com/join/shared_invite/enQtNTk2ODQ1NDU1NDI0LTc3MzQyZmM0ZGFjNzg5MzY1MzJlZTg4YjQ4YTU0ZmMxYWU2MmVlMzhhY2MzMTc1MDg0MzRjNTA4MzRkMGE0Nzk)

Amundsen Metadata service serves Restful API and is responsible for providing and also updating metadata, such as table & column description, and tags. Metadata service can use Neo4j or Apache Atlas as a persistent layer.

For information about Amundsen and our other services, visit the [main repository](https://github.com/amundsen-io/amundsen#amundsen) `README.md`. Please also see our instructions for a [quick start](https://github.com/amundsen-io/amundsen/blob/master/docs/installation.md#bootstrap-a-default-version-of-amundsen-using-docker) setup  of Amundsen with dummy data, and an [overview of the architecture](https://github.com/amundsen-io/amundsen/blob/master/docs/architecture.md#architecture).

## Requirements
- Python >= 3.7

## Doc
- https://www.amundsen.io/amundsen/

## Instructions to start the Metadata service from distribution
```bash
$ venv_path=[path_for_virtual_environment]
$ python3 -m venv $venv_path
$ source $venv_path/bin/activate
$ pip3 install amundsen-metadata
$ python3 metadata_service/metadata_wsgi.py

-- In a different terminal, verify getting HTTP/1.0 200 OK
$ curl -v http://localhost:5002/healthcheck
```

## Instructions to start the Metadata service from the source
```bash
$ git clone https://github.com/amundsen-io/amundsenmetadatalibrary.git
$ cd amundsenmetadatalibrary
$ python3 -m venv venv
$ source venv/bin/activate
$ pip3 install -r requirements.txt
$ python3 setup.py install
$ python3 metadata_service/metadata_wsgi.py

-- In a different terminal, verify getting HTTP/1.0 200 OK
$ curl -v http://localhost:5002/healthcheck
```

## Instructions to start the service from Docker

```bash
$ docker pull amundsendev/amundsen-metadata:latest
$ docker run -p 5002:5002 amundsendev/amundsen-metadata
# - alternative, for production environment with Gunicorn (see its homepage link below)
$ ## docker run -p 5002:5002 amundsendev/amundsen-metadata gunicorn --bind 0.0.0.0:5002 metadata_service.metadata_wsgi

-- In a different terminal, verify getting HTTP/1.0 200 OK
$ curl -v http://localhost:5002/healthcheck
```


## Production environment
By default, Flask comes with Werkzeug webserver, which is for development. For production environment use production grade web server such as [Gunicorn](https://gunicorn.org/ "Gunicorn").

```bash
$ pip install gunicorn
$ gunicorn metadata_service.metadata_wsgi
```
Here is [documentation](https://docs.gunicorn.org/en/latest/run.html "documentation") of gunicorn configuration.

### Configuration outside local environment
By default, Metadata service uses [LocalConfig](https://github.com/amundsen-io/amundsenmetadatalibrary/blob/master/metadata_service/config.py "LocalConfig") that looks for Neo4j running in localhost.
In order to use different end point, you need to create [Config](https://github.com/amundsen-io/amundsenmetadatalibrary/blob/master/metadata_service/config.py "Config") suitable for your use case. Once config class has been created, it can be referenced by [environment variable](https://github.com/amundsen-io/amundsenmetadatalibrary/blob/master/metadata_service/metadata_wsgi.py "environment variable"): `METADATA_SVC_CONFIG_MODULE_CLASS`

For example, in order to have different config for production, you can inherit Config class, create Production config and passing production config class into environment variable. Let's say class name is ProdConfig and it's in metadata_service.config module. then you can set as below:

`METADATA_SVC_CONFIG_MODULE_CLASS=metadata_service.config.ProdConfig`

This way Metadata service will use production config in production environment. For more information on how the configuration is being loaded and used, here's reference from Flask [doc](http://flask.pocoo.org/docs/1.0/config/#development-production "doc").

# Apache Atlas
Amundsen Metadata service can use Apache Atlas as a backend. Some of the benefits of using Apache Atlas instead of Neo4j is that Apache Atlas offers plugins to several services (e.g. Apache Hive, Apache Spark) that allow for push based updates. It also allows to set policies on what metadata is accesible and editable by means of Apache Ranger.

If you would like to use Apache Atlas as a backend for Metadata service you will need to create a [Config](https://github.com/amundsen-io/amundsenmetadatalibrary/blob/master/metadata_service/config.py "Config") as mentioned above. Make sure to include the following:

```python
PROXY_CLIENT = PROXY_CLIENTS['ATLAS'] # or env PROXY_CLIENT='ATLAS'
PROXY_PORT = 21000          # or env PROXY_PORT
PROXY_USER = 'atlasuser'    # or env CREDENTIALS_PROXY_USER
PROXY_PASSWORD = 'password' # or env CREDENTIALS_PROXY_PASSWORD
```

To start the service with Atlas from Docker. Make sure you have `atlasserver` configured in DNS (or docker-compose)

```bash
$ docker run -p 5002:5002 --env PROXY_CLIENT=ATLAS --env PROXY_PORT=21000 --env PROXY_HOST=atlasserver --env CREDENTIALS_PROXY_USER=atlasuser --env CREDENTIALS_PROXY_PASSWORD=password amundsen-metadata:latest
```

---
**NOTE**

The support for Apache Atlas is work in progress. For example, while Apache Atlas supports fine grained access, Amundsen does not support this yet.

# Developer guide
## Code style
- PEP 8: Amundsen Metadata service follows [PEP8 - Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/ "PEP8 - Style Guide for Python Code").
- Typing hints: Amundsen Metadata service also utilizes [Typing hint](https://docs.python.org/3/library/typing.html "Typing hint") for better readability.

## API documentation

We have Swagger documentation setup with OpenApi 3.0.2. This documentation is generated via Flasgger. When adding or updating an API please make sure to update the documentation. To see the documentation run the application locally and go to localhost:5002/apidocs/. Currently the documentation only works with local configuration.

## Code structure
Please visit [Code Structure](docs/structure.md) to read how different modules are structured in Amundsen Metadata service.

## Roundtrip tests
Roundtrip tests are a new feature - by implementing the abstract_proxy_tests and some test setup endpoints in the base_proxy, you can validate your proxy code against the actual data store. These tests do not run by default, but can be run by passing the `--roundtrip-[proxy]` argument. Note this requires
a fully-configured backend to test against.
```bash
$ python -m pytest --roundtrip-neptune .
```


CHANGELOG.md
### Feature
* Add System Statistics Metrics Endpoint ([#228](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/228)) ([`0f47f53`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/0f47f537bad27b6972aeded95200b50125a8243b))
* Implements the table and column lineage for Neo4j ([#294](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/294)) ([`3bc3193`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/3bc3193add5d1772ea1190793414ad0256dd81d7))
* Create users in metadata backend via API ([#289](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/289)) ([`eeba485`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/eeba485b5c211e6d92bbb841bfc26bca79f051e0))
* Add Column Badge API ([#273](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/273)) ([`ee0ac63`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/ee0ac63194a4286715cd32c34042443c979f4cea))
* Column Lineage API ([#280](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/280)) ([`681893f`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/681893fb47ebcfc61ed4736f535618fddb6c4631))
* Table Lineage API ([#262](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/262)) ([`e306034`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/e306034043e067a2d6c24477bd197d10177d1d6d))
* Column badges in Atlas Proxy ([#263](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/263)) ([`a3efb4c`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/a3efb4cb778a6a5a58b0e55874622f114fd2e7e7))
* Added get_lineage method to neo4j proxy ([#259](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/259)) ([`b129cc7`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/b129cc701117871a90ae05c36bcf0f112ae36e5e))
* Neo4j backend for popular tables personalization ([#233](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/233)) ([`d045efa`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/d045efabe116a5cb459441389b814571fb83232b))
* Updated popular_tables endpoint to allow optional user_id ([#232](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/232)) ([`5680775`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/56807753963735eaf1556dae199e0da82b4ebb71))
* AddSwaggerEnabledAsEnvVar ([#215](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/215)) ([`3c9a55e`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/3c9a55e6af4cac9b342803c34cfe81851470e7f5))
* Add neptune proxy ([#204](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/204)) ([`09845d4`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/09845d479297415093033d123868196bd02d19b5))
* Return column level badges ([#205](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/205)) ([`d4d8101`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/d4d81017ab2994375526fb5f4c48092fbb804150))
* Badges with category and badge_type fields ([#201](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/201)) ([`19e1bf8`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/19e1bf8e5c9581ab4b6947f37b7584bac8f2ccf1))
* Get last updated ts for AtlasProxy ([#177](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/177)) ([`3e92586`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/3e9258675ea46beb4192638b4bbebcb20c30cb77))
* Data Owner Implementation of Atlas Proxy ([#156](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/156)) ([`48b4c71`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/48b4c71054f7b34bca8a8f148e8b7499dfb986fc))

### Fix
* Sort reports alphabetically ([#293](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/293)) ([`c8423c6`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/c8423c62c85f94fe4407f2b42dcc176a77849645))
* Reconcile gremlin description writes with the Databuilder ([#290](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/290)) ([`18454fe`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/18454fec95a672e7d01fdbfa91e86df318246457))
* Swagger docs don't align with common entity ([#283](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/283)) ([`db33af9`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/db33af98d6ed2ba29c059b790bc7d09ce71b24c8))
* Compatibility changes  to the gremlin integration ([#260](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/260)) ([`a765424`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/a765424aa13c71afdb04f7457f30d8a8977974a2))
* Proxy client creation fails after adding client_kwargs ([#258](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/258)) ([`1880cec`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/1880cec3bf0ed4b889c5fe3b61633228063cdfad))
* Reapply previous fix ([#245](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/245)) ([`c7dc172`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/c7dc172b2b8c3b8ff4d7f145e5564474a3cb8dd8))
* Expire is a keyword argument for beaker cache ([#239](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/239)) ([`a7b2ec5`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/a7b2ec543470157361767e705f945beed4b32bac))
* Get_tags no longer return tag_count 0 tag ([#230](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/230)) ([`5097d2b`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/5097d2baba2fe8894c15e9627e1ca7bacc3807fb))
* Too many partitions in one Atlas query (Watermarks in Atlas Proxy) ([#217](https://github.com/amundsen-io/amundsenmetadatalibrary/issues/217)) ([`cc3768f`](https://github.com/amundsen-io/amundsenmetadatalibrary/commit/cc3768fb3ae055f0f84958ee5c08560e9c63ffc4))

metadata_service/config.py
# Copyright Contributors to the Amundsen project.
# SPDX-License-Identifier: Apache-2.0

import distutils.util
import os
from typing import Dict, List, Optional, Set  # noqa: F401

import boto3
from amundsen_gremlin.config import LocalGremlinConfig

from metadata_service.entity.badge import Badge

# PROXY configuration keys
PROXY_HOST = 'PROXY_HOST'
PROXY_PORT = 'PROXY_PORT'
PROXY_USER = 'PROXY_USER'
PROXY_PASSWORD = 'PROXY_PASSWORD'
PROXY_ENCRYPTED = 'PROXY_ENCRYPTED'
PROXY_VALIDATE_SSL = 'PROXY_VALIDATE_SSL'
PROXY_CLIENT = 'PROXY_CLIENT'
PROXY_CLIENT_KWARGS = 'PROXY_CLIENT_KWARGS'

PROXY_CLIENTS = {
    'NEO4J': 'metadata_service.proxy.neo4j_proxy.Neo4jProxy',
    'ATLAS': 'metadata_service.proxy.atlas_proxy.AtlasProxy',
    'NEPTUNE': 'metadata_service.proxy.neptune_proxy.NeptuneGremlinProxy'
}

IS_STATSD_ON = 'IS_STATSD_ON'
USER_OTHER_KEYS = 'USER_OTHER_KEYS'


class Config:
    LOG_FORMAT = '%(asctime)s.%(msecs)03d [%(levelname)s] %(module)s.%(funcName)s:%(lineno)d (%(process)d:' \
                 '%(threadName)s) - %(message)s'
    LOG_DATE_FORMAT = '%Y-%m-%dT%H:%M:%S%z'
    LOG_LEVEL = 'INFO'

    # Path to the logging configuration file to be used by `fileConfig()` method
    # https://docs.python.org/3.7/library/logging.config.html#logging.config.fileConfig
    # LOG_CONFIG_FILE = 'metadata_service/logging.conf'
    LOG_CONFIG_FILE = None

    PROXY_USER = os.environ.get('CREDENTIALS_PROXY_USER', 'neo4j')
    PROXY_PASSWORD = os.environ.get('CREDENTIALS_PROXY_PASSWORD', 'test')

    PROXY_ENCRYPTED = True
    """Whether the connection to the proxy should use SSL/TLS encryption."""

    # Prior to enable PROXY_VALIDATE_SSL, you need to configure SSL.
    # https://neo4j.com/docs/operations-manual/current/security/ssl-framework/
    PROXY_VALIDATE_SSL = False
    """Whether the SSL/TLS certificate presented by the user should be validated against the system's trusted CAs."""

    IS_STATSD_ON = False

    # Configurable dictionary to influence format of column statistics displayed in UI
    STATISTICS_FORMAT_SPEC: Dict[str, Dict] = {}

    # whitelist badges
    WHITELIST_BADGES: List[Badge] = []

    SWAGGER_ENABLED = os.environ.get('SWAGGER_ENABLED', False)

    USER_DETAIL_METHOD = None   # type: Optional[function]

    RESOURCE_REPORT_CLIENT = None   # type: Optional[function]

    # On User detail method, these keys will be added into amundsen_common.models.user.User.other_key_values
    USER_OTHER_KEYS = {'mode_user_id'}  # type: Set[str]

    # Number of minimum reader count to qualify for popular table
    POPULAR_TABLE_MINIMUM_READER_COUNT = 10  # type: int

    # List of regexes which will exclude certain parameters from appearing as Programmatic Descriptions
    PROGRAMMATIC_DESCRIPTIONS_EXCLUDE_FILTERS = []  # type: list

    # Custom kwargs that will be passed to proxy client. Can be used to fine-tune parameters like timeout
    # or num of retries
    PROXY_CLIENT_KWARGS: Dict = dict()

    SWAGGER_TEMPLATE_PATH = os.path.join('api', 'swagger_doc', 'template.yml')
    SWAGGER = {
        'openapi': '3.0.2',
        'title': 'Metadata Service',
        'uiversion': 3
    }


class LocalConfig(Config):
    DEBUG = True
    TESTING = False
    LOG_LEVEL = 'DEBUG'
    LOCAL_HOST = '0.0.0.0'

    PROXY_HOST = os.environ.get('PROXY_HOST', f'bolt://{LOCAL_HOST}')
    PROXY_PORT = os.environ.get('PROXY_PORT', 7687)
    PROXY_CLIENT = PROXY_CLIENTS[os.environ.get('PROXY_CLIENT', 'NEO4J')]
    PROXY_ENCRYPTED = bool(distutils.util.strtobool(os.environ.get(PROXY_ENCRYPTED, 'True')))
    PROXY_VALIDATE_SSL = bool(distutils.util.strtobool(os.environ.get(PROXY_VALIDATE_SSL, 'False')))

    IS_STATSD_ON = bool(distutils.util.strtobool(os.environ.get(IS_STATSD_ON, 'False')))

    SWAGGER_ENABLED = True


class AtlasConfig(LocalConfig):
    PROXY_HOST = os.environ.get('PROXY_HOST', 'localhost')
    PROXY_PORT = os.environ.get('PROXY_PORT', '21000')
    PROXY_CLIENT = PROXY_CLIENTS['ATLAS']

    # List of accepted date formats for AtlasProxy Watermarks. With this we allow more than one datetime partition
    # format to be used in tables
    WATERMARK_DATE_FORMATS = ['%Y%m%d']


class GremlinConfig(LocalGremlinConfig, LocalConfig):
    JANUS_GRAPH_URL = None


class NeptuneConfig(LocalGremlinConfig, LocalConfig):
    DEBUG = False
    LOG_LEVEL = 'INFO'

    # PROXY_HOST FORMAT: wss://<NEPTUNE_URL>:<NEPTUNE_PORT>/gremlin
    PROXY_HOST = os.environ.get('PROXY_HOST')
    PROXY_PORT = None   # type: ignore

    PROXY_CLIENT = PROXY_CLIENTS['NEPTUNE']
    PROXY_PASSWORD = boto3.session.Session(region_name=os.environ.get('AWS_REGION', 'us-east-1'))

    PROXY_CLIENT_KWARGS = {
        'neptune_bulk_loader_s3_bucket_name': os.environ.get('S3_BUCKET_NAME'),
        'ignore_neptune_shard': distutils.util.strtobool(os.environ.get('IGNORE_NEPTUNE_SHARD', 'True'))
    }

    JANUS_GRAPH_URL = None
